include/nbl/ui/IWindow.h

```diff
@@ -47,6 +76,8 @@ class IWindow : public core::IReferenceCounted
 
     virtual IClipboardManager* getClipboardManager() = 0;
 
+    virtual bool nextEvent(SEvent* _event) = 0;
```

Comment:
ok this is an evil antipattern

it also ties into input design (how we should handle keyboard, mouse, joystick, etc).

Polling one event at a time is evil, because you need storage for all events anyway.
Callbacks are evil because they require a fuckton of boilerplate to do any "real work" and encourage putting the computation in the callback functions.

I'd much rather expose an event log/queue of <Timestamp,EventType,Event> which you can inspect wholly as you please.

The data structure should hold the following properties:
- Support quick batch removals
- Quick batch finding for events with desired timestamp ranges (or even allow to filter by event type)
- The found batches should be mostly contiguous memory (performance), but they can be segmented
- Be resistant to overflows
- You can assume that events will be pushed onto the queue/log with non-decreasing timestamps (by construction the queue/log will be sorted by timestamp)

So some per-event-type growable Dequeue which supports relatively fast `lower_bound` / `upper_bound`

For example if you used it for Input handling (separate from window events) makes it easy to "spot" keyboard combos such as Ctrl+Shift+V.

Reply:
replace `EventType` with `ChannelID`

we could have the same event type but different channels (imagine multiple gamepads, event types are the same, channels are different.... or imagine actually trying to support multiple keyboards/mice XD)